### PR TITLE
Add experimental support for the new algorithms that stop truncating MAC tags

### PR DESCRIPTION
**Pull request by poljar**
It has become a bit of a pain to rebase this patchset constantly, so I decided that it's a good idea to merge this behind a feature flag.

Depends on #950.

### CODE REVIEW

File: crates/matrix-sdk-crypto/src/gossiping/machine.rs [hunk 1]
```diff
@@ -953,36 +959,53 @@ impl GossipMachine {
...
                     room_id = content.room_id.as_str(),
+                    %algorithm,
                     "Couldn't create a group session from a received room key"
                 );
                 Err(e.into())
             }
         }
     }
 
+    async fn receive_supported_keys(
```
Comment (Hywan) [inline crates/matrix-sdk-crypto/src/gossiping/machine.rs:970]: I understand this is extracted from `receive_forwarded_room_key`, but wouldn't it be great to add a documentation even if the method isn't public, thoughts?

File: crates/matrix-sdk-crypto/src/machine.rs [hunk 2]
```diff
@@ -563,19 +563,37 @@ impl OlmMachine {
...
+            } else {
+                unsupported_warning();
+                Ok(None)
             }
+        };
+
+        match &event.content {
+            RoomKeyContent::MegolmV1AesSha2(content) => handle_key(content),
+            #[cfg(feature = "experimental-algorithms")]
+            RoomKeyContent::MegolmV2AesSha2(content) => handle_key(content),
```
Comment (Hywan) [inline crates/matrix-sdk-crypto/src/machine.rs:596]: I'm surprised `content` is of type `MegolmV1AesSha2Content` here (as `handle_key` will expect), must check.

Comment (Hywan) [reply crates/matrix-sdk-crypto/src/machine.rs:596]: Checked: `MegolmV2AesSha2Content` is a type alias to `MegolmV1AesSha2Content` as defined in https://github.com/matrix-org/matrix-rust-sdk/pull/951/files#diff-e746e9dee91391967b6d55fca252ef46ddbd25f4b0d76e95ddf783094b7779deR138, so it all works :-).

**Review by Hywan — APPROVED** [review]
Excellent work. I'm fine adding it behind a feature flag, it makes sense to avoid the pain of constant rebasing.

File: crates/matrix-sdk-crypto/src/gossiping/machine.rs (hunk 1 above)
Comment (poljar) [reply crates/matrix-sdk-crypto/src/gossiping/machine.rs:970]: Eh, I expect this to move back into `receive_forwarded_room_key` by the time we can remove the feature flag. But sure, I'll add a sentence or two.

**Comment by poljar** [comment]
What happened here? This PR depends on #950 which is waiting for review from @dkasak.

**Pull request merged**